Add nested objects and array support

lucascurti opened the pull request:
@airjp73 I still need to create some tests and work on the documentation updates but the nested objects and arrays should be working. It would be great if you can play a bit with it and let me know what you think. I didn't implemented the option to turn off nesting completely for a field using brackets. I think that one can be added in a later PR.

To run the `sample-app` just run `npm run sample-app` in the root folder.

I'm keeping this one as draft until documentation and tests are added.

airjp73 commented on package.json:
```diff
@@ -14,7 +14,9 @@
-    "test": "jest src",
+    "build:sample": "tsc --project tsconfig.json --module CommonJS --outDir ./sample-app/remix-validated-form",
+    "sample-app": "npm run build:sample && cd sample-app && npm run dev",
+    "test": "jest src --watch",
```
This _might_ break CI, since this is run as part of the github action. We could add a separate `test:watch` script though.

lucascurti replied:
Cool. `test-watch` added.

airjp73 commented on src/validation/validation.test.ts:
```diff
@@ -45,44 +82,54 @@ describe("Validation", () => {
     });
 
     describe("validateField", () => {
       it("should not return an error if field is valid", () => {
```
It could be good to add a test case for `validatedField` that validates a nested field.

lucascurti replied:
tests added

airjp73 reviewed (commented):
Thanks for the work! I booted up the sample app and everything seems to be working correctly. I had a few comments but largely looks good.

airjp73 commented:
Whoops, didn't mean to request a review from myself 😆

lucascurti commented on sample-app/app/components/SubjectForm.tsx:
```diff
@@ -0,0 +1,149 @@
+                  <FormInput
+                    name="description"
+                    label="Description"
+                    isRequired
```
Noticed this too. I think it has to do with Chakra adding `required` to the fields. Adding `noValidate` to the `<form>` fixes the issue by disabling native validation and relying only on `remix-validated-form`. We can investigate this further in #5 and see if we can come up with a better solution.

lucascurti commented on sample-app/.env:
```diff
@@ -0,0 +1,7 @@
+# Prisma supports the native connection string format for PostgreSQL, MySQL, SQLite, SQL Server and MongoDB (Preview).
+# See the documentation for all the connection string options: https://pris.ly/d/connection-strings
+
+DATABASE_URL="file:./dev.db"
```
Me neither! I learned about it from the Jokes Sample App for Remix from Kent C Dodds.

lucascurti commented on package.json:
```diff
@@ -35,43 +37,45 @@
     "zod": "^3.11.6"
   },
   "dependencies": {
+    "lodash": "^4.17.21",
```
The thing is that the `flatten` and `unflatten` functions depends on several functions from `lodash`. I will change to single imports for each function to avoid bundling the entire `lodash` library. This will have the exact same impact than copying the `lodash` implementations to the code.

lucascurti commented on package.json:
```diff
@@ -35,43 +37,45 @@
   },
   "lint-staged": {
-    "**/*.{ts,tsx,js,jsx}": [
+    "**/!(test-app|sample-app)/*.{ts,tsx,js,jsx}": [
```
I was having problems committing the sample-app. Was a temporary fix and I forgot to revert it. It's fixed now.

lucascurti commented on src/flatten.ts:
```diff
@@ -0,0 +1,60 @@
+  return reduce(
+    params,
+    function (result, value, key) {
+      return set(result, key, value);
```
Fixed by importing single functions. See https://github.com/airjp73/remix-validated-form/pull/4/commits/ea3691f33f35a4e4cef6b23cc270275a47b150b5

lucascurti commented:
@airjp73. Thanks for those comments! I implemented them and marked them as resolved. Check my replies and feel free to let me know if you have any other feedback. I'll work in the documentation and let you know as soon as it's ready.

airjp73 commented:
It's looking good to me. :+1:

lucascurti commented:
@airjp73 This one is now ready for review.

- I've updated the README:
  - added info on nested objects and arrays, 
  - uploaded a new Demo video from the `sample-app`
  - added instructions on how to run the `sample-app`
  - added information on `createValidator` 
  - added FAQs section for the `noValidate` option.
- I upgraded to EsLint 8 but it's having some issues when linting so I downgraded to v7 (the one you were originally using) and everything is good now. You might want to run those actions again and see if they pass now.

This PR is now longer in Draft so feel free to merge it when you are ready. 🥳

airjp73 commented:
Looks good, thanks!

airjp73 commented:
@lucascurti, I just published the changes under 1.1.1-beta.0 in case you wanted to use it right away. I'm gonna work on the other issues I made before releasing so I can bump the major version.

lucascurti commented:
Cool! I'll update to the beta tomorrow. 
BTW, in case you find it useful, I recommend https://intuit.github.io/auto/ for publishing to npm, creating release notes, beta/next versioning and a bunch of other things. It's pretty easy to setup.

airjp73 commented:
I'll check it out :+1:
